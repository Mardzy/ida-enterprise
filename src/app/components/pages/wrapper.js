import React from "react";
import { Switch, Route } from "react-router";

import ContentWithSidebar from "../../../ui/layout/content-with-sidebar";

import Chart from "../../../modules/chart";
import Dashboard from "../../../modules/chart-group";
import MainMenu from "../../../modules/navigation";

/**
 * Render a chart.
 * @param {*} chartId
 * @return {*}
 */
const renderChart = chartId => <Chart key={chartId} id={chartId}/>;

/**
 * Render nested routes.
 * @return {*}
 */
const renderContent = () => (
  <Switch>
    <Route render={() => <Dashboard pageId="dashboard" renderChart={renderChart}/>}/>
  </Switch>
);

/**
 * Render main menu on sidebar.
 * @return {*}
 */
const renderSidebar = () => <MainMenu/>;

/**
 * Page wrapper with navigation.
 * @type {ReactComponent}
 */
const WrapperWithNavigation = () => <ContentWithSidebar padded sidebar={renderSidebar} content={renderContent}/>;

export default WrapperWithNavigation;
