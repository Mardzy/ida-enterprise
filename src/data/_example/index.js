import reducer from "./reducers";

export { default as duckEpics } from "./epics";
export { default as duckTypes } from "./types";

export default reducer;