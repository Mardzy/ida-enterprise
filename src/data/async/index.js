import { async as reducer } from "./reducers";

export default reducer;