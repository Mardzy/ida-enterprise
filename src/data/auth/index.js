import authReducer from "./reducers";

export { default as authEpics } from "./epics";
export { default as authTypes } from "./types";
export { default as authActions } from "./actions";

export default authReducer;