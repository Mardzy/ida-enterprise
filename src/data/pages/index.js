import pagesReducer from "./reducers";

export { default as pageEpics } from "./epics";
export { default as pageTypes } from "./types";

export default pagesReducer;