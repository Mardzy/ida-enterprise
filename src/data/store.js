import { fromJS } from "immutable";
import { createEpicMiddleware } from "redux-observable";
import { applyMiddleware, createStore } from "redux";
import { composeWithDevTools } from "redux-devtools-extension";
import persistState from "redux-localstorage";
import { routerMiddleware } from "connected-react-router/immutable";

import { INITIAL_STATE } from "./initial-state";
import { rootReducer, history } from "./reducer";
import { rootEpic } from "./epic";
// import { gaMiddleware } from "data/config/analytics";
import { serialize, deserialize, merge, slice } from "./config/persistent-store";

/**
 * Construct an epic middleware.
 * @type {EpicMiddleware<Action, any, any, Action>}
 */
const epicMiddleware = createEpicMiddleware();

/**
 * Middleware to apply in the store.
 * @type {*[]}
 */
const middleware = [
  // gaMiddleware,
  epicMiddleware,
  routerMiddleware(history)
];

/**
 * Configuration for keeping all the redux state in localstorage.
 * @type {{key: string, slicer: slice, serialize: serialize, deserialize: deserialize, merge: merge}}
 */
const persistStateConfig = {
  key: "partner-dashboard",
  slicer: slice,
  serialize, deserialize, merge
};

/**
 * Create a Redux store.
 * @param {Object} initialState
 * @return {Store<any>}
 */
const configureStore = (initialState = {}) => createStore(
  rootReducer,
  INITIAL_STATE.merge(fromJS(initialState)),
  composeWithDevTools({})(
    applyMiddleware(...middleware),
    persistState(["auth", "farms"], persistStateConfig)
  )
);

/**
 * Create a store with the reducers.
 */
const store = configureStore();

epicMiddleware.run(rootEpic);

export default store;
export { history, epicMiddleware, configureStore };