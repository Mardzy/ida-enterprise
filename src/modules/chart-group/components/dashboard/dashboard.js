import React from "react";
import PT from "prop-types";
import IPT from "react-immutable-proptypes";
import { fromJS } from "immutable";

import MainWithSecondary from "../../../../ui/layout/main-with-secondary";
import SingleColumn from "../../../../ui/layout/single-column";
import DoubleColumn from "../../../../ui/layout/double-column";

/**
 * Section names
 * @memberOf module:Constants
 * @type {{main: string, left: string, right: string}}
 */
export const CHART_GROUP_SECTIONS = {
  main: "main",
  left: "left",
  right: "right"
};

/**
 * Dashboard page with main chart on top, supporting charts on bottom.
 * @type {ReactComponent}
 */
class Dashboard extends React.Component {

  /**
   * @memberOf Dashboard
   * @type {{fetchData: function, sections: Map, customerId: number, pageId: *}}
   */
  static propTypes = {
    /** fetch content */
    fetchData: PT.func,
    /** charts (id, config, etc.) to display in the layout */
    sections: IPT.map,
    /** customer id */
    customerId: PT.number,
    /** page id */
    pageId: PT.any,
    /** chart renderer */
    renderChart: PT.func
  };

  /**
   * @memberOf Dashboard
   * @type {{chartComponent: function(): null, fetchData: function(): null, sections: any}}
   */
  static defaultProps = {
    fetchData: () => null,
    renderChart: () => null,
    sections: fromJS({
      [CHART_GROUP_SECTIONS.main]: [],
      [CHART_GROUP_SECTIONS.left]: [],
      [CHART_GROUP_SECTIONS.right]: []
    })
  };

  /**
   * Render the main section of the layout
   * @memberOf Dashboard
   * @return {*}
   */
  renderMain () {
    return (
      <SingleColumn content={this.renderSection(CHART_GROUP_SECTIONS.main)}/>
    );
  }

  /**
   * Render the secondary section of the layout
   * @memberOf Dashboard
   * @return {*}
   */
  renderSecondary () {
    return (
      <DoubleColumn left={this.renderSection(CHART_GROUP_SECTIONS.left)} right={this.renderSection(CHART_GROUP_SECTIONS.right)}/>
    );
  }

  /**
   * Render a section
   * @memberOf Dashboard
   * @param key
   * @return {function()}
   */
  renderSection (key) {
    return () => this.props.sections.get(key).map(chart => this.props.renderChart(chart));
  }

  render () {
    return <MainWithSecondary main={this.renderMain.bind(this)} secondary={this.renderSecondary.bind(this)}/>;
  }

}

export default Dashboard;