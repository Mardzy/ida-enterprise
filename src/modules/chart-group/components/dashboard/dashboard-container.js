import React from "react";
import { connect } from "react-redux";
import { compose, branch, renderComponent, createSink, lifecycle } from "recompose";

import SpinnerWithOverlay from "../../../../ui/loaders/spinner";
import actions from "../../../../data/pages/actions";
import { authTypes } from "../../../../data/auth";
import { pageTypes } from "../../../../data/pages";
import { actionsResolved } from "../../../../data/async/utils";

import Dashboard from "./dashboard";

/**
 * Tell whether the user info and the chart list is loaded.
 * @param state
 * @param ownProps
 * @return {boolean}
 */
const mayRender = (state, ownProps) =>
  actionsResolved(state, [authTypes.REQUEST_USER, { type: pageTypes.FETCH_CHART_LIST, unique: ownProps.pageId }]);

/**
 * Render a spinner with text.
 * @return {*}
 * @constructor
 */
const SpinnerWithText = () => <SpinnerWithOverlay>Loading charts</SpinnerWithOverlay>;

export default compose(
  /**
   * Connect the component to Redux.
   */
  connect((state, ownProps) => ({
    /**
     * @type {Immutable.Map}
     */
    sections: state.getIn(["pages", ownProps.pageId]),
    /**
     * @type {number}
     */
    customerId: state.getIn(["auth", "user", "customerID"]),
    /**
     * @type {boolean}
     */
    isLoading: !mayRender(state, ownProps)
  }), dispatch => ({
    /**
     * Fetch page data.
     * @param customerId
     * @param pageId
     * @return {*}
     */
    fetchData: (customerId, pageId) => dispatch(actions.fetchChartList.start(customerId, pageId))
  })),
  /**
   * Add life cycle methods externally for better unit testing.
   */
  lifecycle({
    /**
     * Fetch data when the customer id is for the first time available.
     * @param nextProps
     */
    componentWillReceiveProps (nextProps) {
      if (!this.props.customerId && nextProps.customerId) {
        this.props.fetchData(nextProps.customerId.customerId, this.props.pageId);
      }
    }
  }),
  /**
   * Render a spinner with text while the data is being fetched.
   */
  branch(props => props.isLoading, renderComponent(SpinnerWithText), renderComponent(Dashboard))
)(createSink());