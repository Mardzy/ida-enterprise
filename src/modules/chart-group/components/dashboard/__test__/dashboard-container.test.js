import { Provider } from "react-redux";

import store from "../../../../../data/store";

import DashboardContainer from "../dashboard-container";
import { CHART_GROUP_SECTIONS } from "../dashboard";
import { authTypes } from "../../../../../data/auth";
import { pageTypes } from "../../../../../data/pages";

describe("The chart group dashboard container", () => {

  it("renders a spinner while loading", () => {
    const wrapper = mount(
      <Provider store={store}>
        <DashboardContainer/>
      </Provider>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("renders the section contents from the store", () => {
    /**
     * @type {string}
     */
    const pageId = "dashboard";

    /**
     * @param chartId
     */
    const renderChart = chartId => expect(typeof chartId).toBe("number");

    /**
     * @type {{}}
     */
    const sections = {
      [CHART_GROUP_SECTIONS.main]: [{ id: 1, config: {} }],
      [CHART_GROUP_SECTIONS.left]: [{ id: 2, config: {} }],
      [CHART_GROUP_SECTIONS.right]: [{ id: 3, config: {} }]
    };

    store.dispatch({
      type: authTypes.REQUEST_USER.FULFILLED,
      payload: {
        givenName: "Sandor",
        surName: "Kocsis"
      }
    });

    store.dispatch({
      type: pageTypes.FETCH_CHART_LIST.FULFILLED,
      payload: { pageId, layout: sections }
    });

    const wrapper = mount(
      <Provider store={store}>
        <DashboardContainer renderChart={renderChart} pageId={pageId}/>
      </Provider>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  })

});