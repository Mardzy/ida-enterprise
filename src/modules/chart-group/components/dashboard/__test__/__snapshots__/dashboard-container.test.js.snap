// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The chart group dashboard container renders a spinner while loading 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(lifecycle(branch(Sink)))>
    <lifecycle(branch(Sink))
      fetchData={[Function]}
      isLoading={true}
    >
      <branch(Sink)
        fetchData={[Function]}
        isLoading={true}
      >
        <renderComponent(SpinnerWithText)
          fetchData={[Function]}
          isLoading={true}
        >
          <SpinnerWithText
            fetchData={[Function]}
            isLoading={true}
          >
            <SpinnerWithOverlay>
              <Dimmer
                active={true}
                inverted={true}
              >
                <DimmerInner
                  active={true}
                  inverted={true}
                >
                  <Ref
                    innerRef={
                      Object {
                        "current": <div
                          class="ui active transition visible inverted dimmer"
                          style="display: flex;"
                        >
                          <div
                            class="content"
                          >
                            <div
                              class="ui inverted text loader"
                            >
                              Loading charts
                            </div>
                          </div>
                        </div>,
                      }
                    }
                  >
                    <RefFindNode
                      innerRef={
                        Object {
                          "current": <div
                            class="ui active transition visible inverted dimmer"
                            style="display: flex;"
                          >
                            <div
                              class="content"
                            >
                              <div
                                class="ui inverted text loader"
                              >
                                Loading charts
                              </div>
                            </div>
                          </div>,
                        }
                      }
                    >
                      <div
                        className="ui active transition visible inverted dimmer"
                        onClick={[Function]}
                      >
                        <div
                          className="content"
                        >
                          <Loader
                            inverted={true}
                          >
                            <div
                              className="ui inverted text loader"
                            >
                              Loading charts
                            </div>
                          </Loader>
                        </div>
                      </div>
                    </RefFindNode>
                  </Ref>
                </DimmerInner>
              </Dimmer>
            </SpinnerWithOverlay>
          </SpinnerWithText>
        </renderComponent(SpinnerWithText)>
      </branch(Sink)>
    </lifecycle(branch(Sink))>
  </Connect(lifecycle(branch(Sink)))>
</Provider>
`;

exports[`The chart group dashboard container renders the section contents from the store 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(lifecycle(branch(Sink)))
    pageId="dashboard"
    renderChart={[Function]}
  >
    <lifecycle(branch(Sink))
      fetchData={[Function]}
      isLoading={false}
      pageId="dashboard"
      renderChart={[Function]}
      sections={
        Immutable.Map {
          "main": Immutable.List [
            1,
          ],
          "left": Immutable.List [
            2,
          ],
          "right": Immutable.List [
            3,
          ],
        }
      }
    >
      <branch(Sink)
        fetchData={[Function]}
        isLoading={false}
        pageId="dashboard"
        renderChart={[Function]}
        sections={
          Immutable.Map {
            "main": Immutable.List [
              1,
            ],
            "left": Immutable.List [
              2,
            ],
            "right": Immutable.List [
              3,
            ],
          }
        }
      >
        <renderComponent(Dashboard)
          fetchData={[Function]}
          isLoading={false}
          pageId="dashboard"
          renderChart={[Function]}
          sections={
            Immutable.Map {
              "main": Immutable.List [
                1,
              ],
              "left": Immutable.List [
                2,
              ],
              "right": Immutable.List [
                3,
              ],
            }
          }
        >
          <Dashboard
            fetchData={[Function]}
            isLoading={false}
            pageId="dashboard"
            renderChart={[Function]}
            sections={
              Immutable.Map {
                "main": Immutable.List [
                  1,
                ],
                "left": Immutable.List [
                  2,
                ],
                "right": Immutable.List [
                  3,
                ],
              }
            }
          >
            <MainWithSecondary
              main={[Function]}
              secondary={[Function]}
            >
              <SingleColumn
                content={[Function]}
              >
                <Grid
                  columns={1}
                  stackable={true}
                >
                  <div
                    className="ui stackable one column grid"
                  >
                    <GridColumn>
                      <div
                        className="column"
                      >
                        <Component />
                      </div>
                    </GridColumn>
                  </div>
                </Grid>
              </SingleColumn>
              <DoubleColumn
                left={[Function]}
                right={[Function]}
              >
                <Grid
                  columns={2}
                  stackable={true}
                >
                  <div
                    className="ui stackable two column grid"
                  >
                    <GridColumn
                      computer={8}
                      tablet={16}
                    >
                      <div
                        className="eight wide computer sixteen wide tablet column"
                      >
                        <Component />
                      </div>
                    </GridColumn>
                    <GridColumn
                      computer={8}
                      tablet={16}
                    >
                      <div
                        className="eight wide computer sixteen wide tablet column"
                      >
                        <Component />
                      </div>
                    </GridColumn>
                  </div>
                </Grid>
              </DoubleColumn>
            </MainWithSecondary>
          </Dashboard>
        </renderComponent(Dashboard)>
      </branch(Sink)>
    </lifecycle(branch(Sink))>
  </Connect(lifecycle(branch(Sink)))>
</Provider>
`;
